ui/pages/confirmations/utils/dapp-swap-comparison-utils.ts

```diff
@@ -0,0 +1,221 @@
+} from '@metamask/transaction-controller';
+import { getNativeTokenAddress } from '@metamask/assets-controllers';
+
+export const ABI = [
+  {
+    constant: true,
+    inputs: [
+      {
+        name: 'commands',
+        type: 'bytes',
+      },
+      {
+        name: 'inputs',
+        type: 'bytes[]',
+      },
+      {
+        name: 'deadline',
+        type: 'uint256',
+      },
+    ],
+    name: 'execute',
+    type: 'function',
+  },
+  {
+    constant: true,
+    inputs: [
+      {
+        name: 'commands',
+        type: 'bytes',
+      },
+      {
+        name: 'inputs',
+        type: 'bytes[]',
+      },
+    ],
+    name: 'execute',
+    type: 'function',
+  },
+];
+
+const COMMAND_BYTE_SWEEP = '04';
+const COMMAND_BYTE_SEAPORT = '10';
+const COMMAND_BYTE_UNWRAP_WETH = '0c';
+
+function getArgsFromInput(input: string) {
+  return input?.slice(2).match(/.{1,64}/gu) ?? [];
+}
+
+function argToAddress(arg?: string) {
+  if (!arg) {
+    return undefined;
+  }
+  return addHexPrefix(arg?.slice(24));
+}
+
+function argToAmount(arg?: string) {
+  if (!arg) {
+    return undefined;
+  }
+  const amount = arg?.replace(/^0+/u, '') || '0';
+  return addHexPrefix(amount);
+}
+
+function parseTransactionData(data?: string) {
+  const contractInterface = new Interface(ABI);
+
+  let parsedTransactionData: TransactionDescription;
+
+  try {
+    parsedTransactionData = contractInterface.parseTransaction({
+      data: data as Hex,
+    });
+  } catch (error) {
+    return { inputs: [], commandBytes: [] };
+  }
+
+  const { commands } = parsedTransactionData.args;
+  const { inputs } = parsedTransactionData.args;
+  const commandBytes = commands.slice(2).match(/.{1,2}/gu) ?? [];
+
+  return { inputs, commandBytes };
+}
+
+export function getDataFromSwap(chainId: Hex, data?: string) {
+  let amountMin;
+  const tokenAddresses = [];
+  const { commandBytes, inputs } = parseTransactionData(data);
+
+  const seaportIndex = commandBytes.findIndex(
+    (commandByte: string) => commandByte === COMMAND_BYTE_SEAPORT,
+  );
+  let seaportArgs: string[] = [];
+  let sweepArgs: string[] = [];
+  let unwrapWethArgs: string[] = [];
+
+  if (seaportIndex >= 0) {
+    seaportArgs = getArgsFromInput(inputs[seaportIndex]);
+    amountMin = argToAmount(seaportArgs[13]);
+  } else {
+    return { quotesInput: undefined, amountMin: undefined, tokenAddresses: [] };
+  }
+
+  const sweepIndex = commandBytes.findIndex(
+    (commandByte: string) => commandByte === COMMAND_BYTE_SWEEP,
+  );
+  if (sweepIndex >= 0) {
+    sweepArgs = getArgsFromInput(inputs[sweepIndex]);
+    amountMin = argToAmount(sweepArgs[2]);
+  }
+
+  const unwrapWethIndex = commandBytes.findIndex(
+    (commandByte: string) => commandByte === COMMAND_BYTE_UNWRAP_WETH,
+  );
+  if (unwrapWethIndex >= 0) {
+    unwrapWethArgs = getArgsFromInput(inputs[unwrapWethIndex]);
+    amountMin = argToAmount(unwrapWethArgs[1]);
+  }
+
+  tokenAddresses.push(argToAddress(seaportArgs[10]));
```

Comment:
Minor, still lots of inline indexes for specific args, could benefit from `parseSeaport` and `decodeSeaportCommand` in future to return readable objects such as `{ sourceTokenAddress, targetTokenAddress }` etc.